crates/sui-package-resolver/src/cache.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::store::{DbPackageStore, PackageStore};
+use crate::Result;
+use crate::{Package, ResolutionContext};
+use lru::LruCache;
+use move_core_types::account_address::AccountAddress;
+use move_core_types::language_storage::TypeTag;
+use move_core_types::value::MoveTypeLayout;
+use std::num::NonZeroUsize;
+use std::sync::{Arc, Mutex};
+use sui_indexer::indexer_reader::IndexerReader;
+use sui_types::is_system_package;
+
+const PACKAGE_CACHE_SIZE: NonZeroUsize = unsafe { NonZeroUsize::new_unchecked(1024) };
+
+/// Cache to answer queries that depend on information from move packages: listing a package's
+/// modules, a module's structs and functions, the definitions or layouts of types, etc.
+///
+/// Queries that cannot be answered by the cache are served by loading the relevant package as an
+/// object and parsing its contents.
+pub struct PackageCache {
+    packages: Mutex<LruCache<AccountAddress, Arc<Package>>>,
+    store: Box<dyn PackageStore + Send + Sync>,
```

Comment:
@amnn Was there a particular reason for making this a trait object vs a generic?

Reply:
I can't think of a reason, and I'm not opposed to this being made generic -- I don't think I tried that in this case.

Reply:
I don't think we need to change it, I was just asking because if this was generic we could add functions to get a reference to the internal store so that different users could call type specific methods on the internal store, when the concrete type is know in a context. This could prevent the need to hold onto a second handle to the rocksdb database to do updates in the case of the analytics pipeline.